### PR TITLE
✨(frontend) split dashboard certifiate page

### PR DESCRIPTION
**Pull request by rlecellier**
## Purpose

joanie's certificate endpoint now return only order's certificate by
default.
to retrive enrollments certificate we need a second request with the
right filter.
Here we add tab in learner dashboard certificate page to render both
list: order and enrollments certificates

image
[Capture vidéo du 25-04-2024 16:39:29.webm](https://github.com/openfun/richie/assets/139848/de8bb7d2-a448-465e-a70c-5886ff0b000b)

### CODE REVIEW

File: src/frontend/js/components/Tabs/_styles.scss [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+//       &--active {
+//         border-bottom: 2px solid $background-color;
+//         margin-bottom: -2px;
+//       }
+//     }
+
+//     margin-bottom: rem-calc(16px);
+//   }
+// }
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/_styles.scss:29]: Debug residue

File: src/frontend/js/components/Tabs/_styles.scss [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  }
+}
+
+.c__button {
+  $default-color: map-get($button-tab-scheme, 'default-color');
+
+  &--tab {
+    // avoid movement when hovering/selecting a tab.
+    border: 1px solid transparent;
+    border-bottom-width: 2px;
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/_styles.scss:66]:
```suggestion
    border-bottom: 2px solid transparent;
```

File: src/frontend/js/components/Tabs/_styles.scss [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  'tab'
+);
+$default-border-radius: map-get($button-tab-scheme, 'border-radius');
+$default-color: map-get($button-tab-scheme, 'default-color');
+$hover-color: map-get($button-tab-scheme, 'hover-color');
+$focus-color: map-get($button-tab-scheme, 'focus-color');
+$selected-color: map-get($button-tab-scheme, 'selected-color');
+$disabled-color: map-get($button-tab-scheme, 'disabled-color');
+
+$tabs-scheme: map-get(map-get(map-get($themes, 'default'), 'components'), 'tabs');
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/_styles.scss:42]: This should work
```suggestion
$tabs-scheme: map-get($themes, 'default', 'components', 'tabs');

$button-tab-scheme: map-get($themes, 'default', 'components', 'button', 'tab');
$default-border-radius: map-get($button-tab-scheme, 'border-radius');
$default-color: map-get($button-tab-scheme, 'default-color');
$hover-color: map-get($button-tab-scheme, 'hover-color');
$focus-color: map-get($button-tab-scheme, 'focus-color');
$selected-color: map-get($button-tab-scheme, 'selected-color');
$disabled-color: map-get($button-tab-scheme, 'disabled-color');
```

File: src/frontend/cunningham.cjs [hunk 4]
```diff
@@ -65,8 +65,18 @@ module.exports = {
...
+        },
         button: {
           'font-family': 'Montserrat',
+          tab: {
+            'default-color': '#555F6B', // --c--theme--colors--greyscale-700
+            'hover-color': '#0C1A2B', // --c--theme--colors--greyscale-900
+            'selected-color': '#0556BF', // --c--theme--colors--primary-600
+            'focus-color': '#555F6B', // --c--theme--colors--greyscale-300
+            'disabled-color': '#79818A', // --c--theme--colors--greyscale-600
+          },
```
Comment (jbpenrath) [inline src/frontend/cunningham.cjs:79]: I do not see tab as a variant of button component. I'll put this property at the root level of `components`.

File: src/frontend/js/components/Tabs/_styles.scss [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    // avoid movement when hovering/selecting a tab.
+    border: 1px solid transparent;
+    border-bottom-width: 2px;
+
+    padding-left: 0;
+    padding-right: 0;
+    color: $default-color;
+    background-color: transparent;
+    border-bottom-right-radius: 0;
+    border-bottom-left-radius: 0;
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/_styles.scss:73]: I do not see where you are applying `border-radius`.

File: src/frontend/js/components/Tabs/_styles.scss [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    border-bottom: 2px solid map-get($tabs-scheme, 'border-bottom-color');
+
+    &__tab {
+      margin-bottom: rem-calc(-2px);
+      flex: 0 0 auto;
+
+      margin-right: rem-calc(16px);
+      &:last-child {
+        margin-right: 0;
+      }
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/_styles.scss:55]: We are in a flex context, I'll use `gap` instead.

File: src/frontend/js/components/Tabs/index.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  const isActive = activeTabName ? activeTabName === name : initialIsActive;
+
+  const handleOnClick = () => {
+    setActiveTabName?.(name);
+    onClick?.();
+    if (href) {
+      navigate(href);
+    }
+  };
+  // className={classNames('tabs__header__tab', { 'tabs__header__tab--active': isActive })}
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/index.tsx:62]: Debug residue.

File: src/frontend/js/components/Tabs/index.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+Tabs.Content = ({ children }: PropsWithChildren) => {
+  return <div className="tabs__content">{children}</div>;
+};
+
+interface TabProps extends PropsWithChildren {
+  name: string;
+  href?: string;
+  onClick?: () => void;
+  initialIsActive?: boolean;
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/index.tsx:46]: IMO, to set the initial active tab, we should do it at the context level not here.

File: src/frontend/js/components/Tabs/TabsContext/index.tsx
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/TabsContext/index.tsx:1]: This file is empty. Just a little mess in commit organization ?

File: src/frontend/scss/vendors/cunningham-tokens.scss [hunk 9]
```diff
@@ -138,9 +138,9 @@ $themes: (
         'st': 0.25rem,
       ),
       'transitions': (
-        'ease-in': cubic-bezier(0.32, 0, 0.67, 0),
-        'ease-out': cubic-bezier(0.33, 1, 0.68, 1),
-        'ease-in-out': cubic-bezier(0.65, 0, 0.35, 1),
+        'ease-in': #{cubic-bezier(0.32, 0, 0.67, 0)},
+        'ease-out': #{cubic-bezier(0.33, 1, 0.68, 1)},
+        'ease-in-out': #{cubic-bezier(0.65, 0, 0.35, 1)},
```
Comment (jbpenrath) [inline src/frontend/scss/vendors/cunningham-tokens.scss:143]: IMO you make a change then revert it and forgot to regenerate tokens

File: CHANGELOG.md [hunk 10]
```diff
@@ -12,6 +12,8 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Added
 
+- Add a tab in learner dashboard certificate page to render both order
+  certificate and enrollment certificate lists.
```
Comment (jbpenrath) [inline CHANGELOG.md:16]: Move it in Unreleased section

File: src/frontend/js/pages/DashboardCertificates/index.tsx [hunk 11]
```diff
@@ -1,74 +1,51 @@
...
   },
-  empty: {
-    defaultMessage: 'You have no certificates yet.',
-    description: 'Message displayed when there are no certificates',
-    id: 'components.DashboardCertificates.empty',
+  enrollmentCertificateTabLabel: {
+    id: 'components.DashboardCertificates.enrollmentCertificateTabLabel',
+    description: 'The label of the enrollment certificate tab',
+    // FIXME(rlecellier): we need a better name for thoses
+    defaultMessage: 'Enrollment certificates',
```
Comment (jbpenrath) [inline src/frontend/js/pages/DashboardCertificates/index.tsx:18]:
```suggestion
    defaultMessage: 'Attestation of achievement',
```

File: src/frontend/js/widgets/Dashboard/utils/learnerRoutesPaths.tsx [hunk 12]
```diff
@@ -42,6 +44,16 @@ export const LEARNER_DASHBOARD_ROUTE_LABELS = defineMessages<LearnerDashboardPat
...
+  [LearnerDashboardPaths.ORDER_CERTIFICATES]: {
+    id: 'components.Dashboard.DashboardRoutes.certificates.order.label',
+    description: 'Label of the order certificates view used in navigation components.',
+    defaultMessage: 'My certificates',
+  },
+  [LearnerDashboardPaths.ENROLLMENT_CERTIFICATES]: {
+    id: 'components.Dashboard.DashboardRoutes.certificates.enrollment.label',
+    description: 'Label of the enrollment certificates view used in navigation components.',
+    defaultMessage: 'My enrollment certificates',
+  },
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/utils/learnerRoutesPaths.tsx:56]: IMO There should be only one entry the sidebar menu `My certificates`. Once on this view, the user should be able to switch between Certificates and Attestation of achievement.

File: src/frontend/js/components/Tabs/_styles.scss (hunk 2 above)
Comment (rlecellier) [reply src/frontend/js/components/Tabs/_styles.scss:66]: you right it work, that was a residue from my previous design, this fix isn't needed anymore

File: src/frontend/js/widgets/Dashboard/utils/learnerRoutesPaths.tsx (hunk 12 above)
Comment (rlecellier) [reply src/frontend/js/widgets/Dashboard/utils/learnerRoutesPaths.tsx:56]: That's the case, the sidebar link to certificate with the default tab "certificate".
Here, we provide a link to be able to access this menu section on the certificate tab or on the attestation tab.

File: src/frontend/js/components/Tabs/index.tsx (hunk 8 above)
Comment (rlecellier) [reply src/frontend/js/components/Tabs/index.tsx:46]: I didn't feel the need to create an custom provider, it's feel overkill to me but i'll do it.

File: src/frontend/js/components/Tabs/_styles.scss (hunk 5 above)
Comment (rlecellier) [reply src/frontend/js/components/Tabs/_styles.scss:73]: Cunningham's button already have border radius all around.
I removed the bottom one to fit with the tab's underline.

Comment (jbpenrath) [reply src/frontend/js/components/Tabs/_styles.scss:33]: I wonder if we could not use button text variant ?

File: src/frontend/cunningham.cjs (hunk 4 above)
Comment (jbpenrath) [reply src/frontend/cunningham.cjs:79]: In styles, you are not using button tokens so I'm not quit sure you'll have to copy button tokens.

Except about the border-radius to set it to 0 ? 🤔

File: src/frontend/js/components/Tabs/_styles.scss (hunk 5 above)
Comment (rlecellier) [reply src/frontend/js/components/Tabs/_styles.scss:33]: It's basicaly what this is, the `--tab` button is a override of cunningham's text button.
It's everything the text button is, with ajustement to be a tab button in tabs design.

I can move the class outside of `.c__button` and move the variable outside of `components.button` but at the end, it'll be a cunningham button with some dedicated css.
It's look like a button variant.

Comment (jbpenrath) [reply src/frontend/js/components/Tabs/_styles.scss:33]: There are text variant for buttons in Cunningham. We could use them directly. That's what I meaned.

https://openfun.github.io/cunningham/storybook/?path=/story/components-button--primary-text

File: src/frontend/cunningham.cjs [hunk 13]
```diff
@@ -65,6 +65,9 @@ module.exports = {
         },
       },
       components: {
+        tabs: {
+          'border-bottom-color': '#E7E8EA', // --c--theme--colors--greyscale-300
```
Comment (jbpenrath) [inline src/frontend/cunningham.cjs:69]: This should work 😉 
```suggestion
          'border-bottom-color': 'ref(theme.colors.greyscale-300)', // --c--theme--colors--greyscale-300
```

File: src/frontend/js/components/Tabs/index.tsx [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    onClick?.();
+    if (href) {
+      navigate(href);
+    }
+  };
+
+  return (
+    <div className="tabs__header__tab">
+      <Button
+        color="primary-text"
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/index.tsx:70]: I'll use `tertiary-text` here.
WDYT ?
image

File: src/frontend/js/components/Tabs/_styles.scss [hunk 15]
```diff
@@ -0,0 +1,42 @@
+$tabs-scheme: map-get($themes, 'default', 'components', 'tabs');
+$selected-color: map-get($themes, 'default', 'color', 'primary-600');
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/_styles.scss:2]: By trying locally, I feel this variable as no value. What do you think to define this color through the tabs-scheme ?

File: src/frontend/cunningham.cjs (hunk 13 above)
Comment (jbpenrath) [inline src/frontend/cunningham.cjs:69]: I'll add `selected-color` here :
```
          'selected-color': 'ref(theme.colors.greyscale-800)',
```

File: src/frontend/js/pages/DashboardCertificates/index.tsx [hunk 16]
```diff
@@ -1,74 +1,50 @@
...
+        >
+          <FormattedMessage {...messages.orderCertificateTabLabel} />
+        </Tabs.Tab>
+        <Tabs.Tab
+          name="enrollment-certificate-tab"
+          initialIsActive={certificateType === CertificateType.ENROLLMENT}
+          href={generatePath(LearnerDashboardPaths.ENROLLMENT_CERTIFICATES)}
+        >
+          <FormattedMessage {...messages.enrollmentCertificateTabLabel} />
+        </Tabs.Tab>
```
Comment (jbpenrath) [inline src/frontend/js/pages/DashboardCertificates/index.tsx:42]: I just remember that we talked to hide this tab if the user has no certificate of this kind. Maybe we could do that in another PR ?

Comment (rlecellier) [reply src/frontend/js/pages/DashboardCertificates/index.tsx:42]: i would prefere yes, it's a different journey.

File: src/frontend/js/components/Tabs/index.tsx [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+interface TabsContextProps {
+  activeTabName?: string;
+  setActiveTabName?: Dispatch<SetStateAction<Maybe<string>>>;
+}
+const TabsContext = createContext<TabsContextProps>({});
+
+const TabContextProvider = ({
+  initialActiveTabName,
+  children,
+}: PropsWithChildren & { initialActiveTabName?: string }) => {
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/index.tsx:24]:
```suggestion
}: PropsWithChildren<{ initialActiveTabName?: string }>) => {
```

File: src/frontend/js/components/Tabs/index.tsx [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+    return { activeTabName, setActiveTabName };
+  }, [activeTabName]);
+
+  return <TabsContext.Provider value={contextValue}>{children}</TabsContext.Provider>;
+};
+
+const Tabs = ({
+  initialActiveTabName,
+  children,
+}: PropsWithChildren & { initialActiveTabName: string }) => {
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/index.tsx:36]:
```suggestion
}: PropsWithChildren<{ initialActiveTabName: string }>) => {
```

File: src/frontend/js/components/Tabs/index.tsx [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+  name: string;
+  href?: string;
+  onClick?: () => void;
+}
+
+Tabs.Tab = ({ name, href, onClick, children }: TabProps) => {
+  const { activeTabName, setActiveTabName } = useContext(TabsContext);
+  const navigate = useNavigate();
+
+  const isActive = !!activeTabName && activeTabName === name;
```
Comment (jbpenrath) [inline src/frontend/js/components/Tabs/index.tsx:64]: I think this should work ?
```suggestion
  const isActive = activeTabName === name;
```

File: src/frontend/js/pages/DashboardCertificates/components/CertificateList/index.tsx [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+        <Spinner aria-labelledby="loading-certificates-data">
+          <span id="loading-certificates-data">
+            <FormattedMessage {...messages.loading} />
+          </span>
+        </Spinner>
+      )}
+
+      {certificates.items.length === 0 && (
+        <Banner message={intl.formatMessage(messages.empty)} type={BannerType.INFO} />
+      )}
```
Comment (jbpenrath) [inline src/frontend/js/pages/DashboardCertificates/components/CertificateList/index.tsx:62]: IMO, we should not display this banner until we are fetching data.

**Review by jbpenrath — APPROVED** [review]
Last few comments then it will be good
